chimedb/dataset/get.py

```diff
@@ -21,42 +24,300 @@
 _logger.addHandler(logging.NullHandler())
 
 
-def get_dataset(id):
+# local cache for states etc so they don't get transferred from the database repeatedly
+_state_cache = dict()
+_dataset_cache = dict()
+_type_cache = dict()
+
+
+class DatasetStateType(orm.DatasetStateType):
+    """
+    Model for datasetstatetype table.
+
+    This wraps the model in chimedb.database.orm and adds a local cache and additional
+    functionality.
+    """
+
+    @classmethod
+    def from_name(cls, name):
+        """Create a new DatasetStateType object.
+
+        Get the table row from cache or database.
+        This does not insert a new row into the database.
+
+        Parameter
+        ----------
+        name : str
+            Name of the type.
+
+        Returns
+        -------
+        DatasetStateType
+            The requested type.
+
+        Raises
+        ------
+        chimedb.dataset.orm.DatasetStateType.DoesNotExist
+            If the type with the requested name doesn't exist in the database. Use
+            chimedb.dataset.insert to create rows.
+        """
+        try:
+            # look in cache first
+            return _type_cache[name]
+        except KeyError:
+            # ask database
+            new_type = orm.DatasetStateType.get(orm.DatasetStateType.name == name)
+            if new_type is None:
+                return None
+
+            # fill cache
+            _type_cache[name] = new_type
+            return _type_cache[name]
+
+    def __repr__(self):
+        return "<get.DatasetStateType: {}>".format(self.name)
+
+
+class DatasetState(orm.DatasetState):
```

Comment:
It would be good to add some logic to fetch types from the cache if they exist. At the moment this class will generate a query to fetch the type, and it does it in unexpected places, e.g. `print(dataset_from_cache.state)` will perform a query as it uses the type in the repr. That's unnecessary because your cache already has the type.

Reply:
The problem is that the type is a foreign key and the key is not the name but an ID. To translate from the foreign key to the cache, we have to query it to get the name. We can cache it after it was queried once though...

Reply:
Or we could also cache them by that ID.

Reply:
I don't mind it fetching it from the DB if the object doesn't exist in the cache, but if it does it should at least try to pull from there.

An easy thing to do would be to use two dicts for the cache, they both point at the same objects but one keys by ID and the other by name.